### PR TITLE
Prepare for generalized outcomes

### PR DESCRIPTION
**Pull request by ihincks**
This is a beefy PR addressing issue #66. The nutshell improvement is that we want experiment outcomes to be able to have fancy data types (like expparams), and to be compatible with future improvements that lift the constraint that any experiment must have a finite number of possible outcomes (this constraint comes in the form of some methods like bayes_risk enumurating over all possible outcomes).

**This is handled as follows:**

- new abstract class called `Domains`, not much to it, just tells you stuff like whether it's finite and is able to spit out values in the domain. This comes with obvious subclasses like `IntegerDomain` and `RealDomain`.
- method `domain(self, expparam)` of `Simulateable`, returning a `Domain` for every set of expparms.
- subclass `FiniteOutcomeModel` of `Model` which can have outcome-enumerating methods
- some other helper properties and methods that you don't usually need to worry about (ex: `are_expparam_dtypes_consistent` tells you if all the expparams correspond to the same datatype)

**Handling breaking changes:**

There are two breaking changes, and they're both very easy to fix on existing models. You need to inherit from `FiniteOutcomeModel` instead of `Model`, and you need to implement the `domain` method, returning for example `IntegerDomain(min=0, max=1)` for a two outcome model. Another thing that comes up is that `Model.pr0_to_likelihood_array` is moved to `FiniteOutcomeModel.pr0_to_likelihood_array`.

**New Features**

- Model testing: Most of the lines of code added in this PR come from a new systematic set of tests for models. There is now a set of tests for every single model and derived model defined in qinfer. These tests do not do anything inference theoretic, they mainly test that models are type-consistent, shape-consistent, fully-implemented, and bug free.
- As a proof of principle, I implemented a `MultinomialModel` (which requires fancy outcome data type which is a tuple of results) which generalizes `BinomialModel`.

**TODO**

- I went through all files in qinfer and tried to make sure they were consistent with this PR. To my knowledge, the only remaining things to fix are bugs not caught by testing.
- However, I did not touch the documentation (I did comment code very well).

**NOTES**

I pulled in the feature-improved-docs because it would be confusing for me not to.

### CODE REVIEW

**Comment by scasagrande** [comment]
> 62 files changed
> 95 commits

**Comment by whitewhim2718** [comment]
Great work on the PR Ian. Ian has made fairly reasonable arguments to me
that these changes should be included in V1.0

On Tue, Aug 16, 2016 at 5:53 PM, Steven Casagrande <notifications@github.com

**Comment by cgranade** [comment]
Now that I'm back from travel, trying to catch up on PRs, I apologize for the delay and thank you for your wonderful contribution. ``master`` should now be much more up to date after merging in the other PRs; could you please merge those in? Thank you again! ♥

**Comment by ihincks** [comment]
Thanks Chris! Hopefully this won't be too painful, and gets done today.

**Comment by ihincks** [comment]
Oh, that was actually pretty easy. The tests seem to be failing because qutip isn't being imported in one of the new tomography model tests and so what should be the qutip module is actually `None`. I don't presently know enough about how these tests work to fix this...

This PR should probably also include a guide on outcomes. I was pretty careful going through inline documentation making sure it still made sense, but it's possible some of the examples or guides need to deal with outcomes.

**Comment by ihincks** [comment]
@whitewhim2718 pointed out that in tomography, version `>=3.1` of qutip was being required. However, in version 3.1, required functions like `rand_dm_ginibre` do not exist. Therefore I changed to version `>=3.2`.

I'm not sure if this is the right thing to do because 3.1 is the highest official release of qutip, though it is quite old (2014).

**Comment by cgranade** [comment]
I think depending on QuTiP 3.2 prereleases is fine, since as you point out, it is quite old, and as the official 3.2 is scheduled to come out very soon. In particular, many of the features such as ``rand_dm_ginibre`` were contributed to QuTiP precisely for QInfer tomography support.

**Comment by cgranade** [comment]
After thinking on it for the weekend, I think the best way forward with this PR is for me to ``git cherry-pick`` the QuTiP CI configuration and 3.2 requirements into a separate PR so that I can test them separately. If that works, that will make the rest of the PR much easier to merge in.

File: src/qinfer/tests/test_concrete_models.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    RandomizedBenchmarkingModel,
+    PoisonedModel, BinomialModel, MultinomialModel,
+    MLEModel, RandomWalkModel,
+    NormalDistribution,
+    BetaDistribution, UniformDistribution,
+    PostselectedDistribution,
+    ConstrainedSumDistribution
+)
+from qinfer.ale import ALEApproximateModel
+from qinfer.tomography import TomographyModel, DiffusiveTomographyModel, pauli_basis, GinibreDistribution
```
Comment (cgranade) [inline src/qinfer/tests/test_concrete_models.py]: It should be possible to use [``unittest.skipIf``](https://docs.python.org/2/library/unittest.html#unittest.skipIf) to guard tests which depend on QuTiP, such that they won't be run unless QuTiP is installed and is at least version 3.2.0.

**Comment by ihincks** [comment]
Okay, added a ``skipIf`` to the ``TestTomographyModel``. Thanks for pointing out this feature, @cgranade .

File: src/qinfer/utils.py [hunk 2]
```diff
@@ -46,6 +46,47 @@
...
 
+def get_qutip_module(required_version='3.2'):
+    """
+    Attempts to return the qutip module, but 
+    silently returns ``None`` if it can't be 
+    imported, or doesn't have version at 
+    least ``required_version``.
+
+    :param str required_version: Valid input to 
+        ``distutils.version.LoosVersion``.
```
Comment (cgranade) [inline src/qinfer/utils.py]: Minor typo here, should be ``LooseVersion``.

File: src/qinfer/domains.py [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+    @property
+    def example_point(self):
+        """
+        Returns any single point guaranteed to be in the domain, but 
+        no other guarantees; useful for testing purposes. 
+        This is given as a size 1 ``np.array`` of type ``dtype``.
+
+        :type: ``np.ndarray``
+        """
+        return np.array([([self.n_meas] + list(0 for i in xrange(self.n_elements - 1)),)], dtype=self.dtype)
```
Comment (cgranade) [inline src/qinfer/domains.py]: pylint is warning about this line in the CI builds, is it possible to change to ``range``? That might mandate using future's ``builtins`` module to ensure that ``range`` doesn't generate a large intermediate list.

**Comment by cgranade** [comment]
I think this looks really good! Thanks for everything with this PR. I want to go through in a bit more detail, as well as discuss w/ @csferrie, but I think this is nearly ready to merge in.

File: src/qinfer/domains.py (hunk 3 above)
Comment (ihincks) [inline src/qinfer/domains.py]: Ah, yes. Not sure why I had that there at all for a constant list. And ``n_elements`` refers to the length of the multinomial tuple, and should never be very big.

File: src/qinfer/tomography/models.py [hunk 4]
```diff
@@ -88,6 +88,9 @@ def __init__(self, basis, allow_subnormalized=False):
         self._allow_subnormalied = allow_subnormalized
         super(TomographyModel, self).__init__()
 
+        # TODO: allow more outcomes
+        self._domain = IntegerDomain(min=0, max=0)
```
Comment (cgranade) [inline src/qinfer/tomography/models.py]: Sorry for my not fully understanding yet, but does this contradict the definition of ``n_outcomes`` below? My impression had been that ``FiniteOutcomeModel`` uses ``n_outcomes`` to set ``domain``, but this snippet and L137 below seem to be defining ``domain`` as returning copies of ``IntegerDomain(0, 0)`` (1-outcome models).

**Comment by cgranade** [comment]
Looking in more detail, I had a bit of confusion about the roles of the new properties and methods (in particular, ``domain``) in relation to previous interfaces (e.g.: ``n_outcomes``). Would you be willing to add a bit to ``doc/guide/models.rst`` describing the changed interfaces? If not, I'm also happy to work my way through it and write said docs. Thanks for the huge contribution!

File: src/qinfer/tomography/models.py (hunk 4 above)
Comment (ihincks) [inline src/qinfer/tomography/models.py]: This is a typo, good catch, should be ``IntegerDomain(0,1)``.

**Comment by ihincks** [comment]
The role of ``n_outcomes`` is definitely something that deserves attention. In this PR, ``n_outcomes`` is completely independent of ``domain``. I will explain in detail later. In short, part of the problem is wanting to keep a property called ``n_outcomes`` in the new system that means for ``FiniteOutcomeModel`` what it did for ``Model`` before this PR. For finite-outcome models, it will typically hold that ``n_outcomes(expparam)=domain(expparam).n_members``. For models with an infinite number of outcomes (or a very large number of outcomes such that enumerating them is painful), ``n_members`` is not defined, but ``n_outcomes`` sort of refers to "enough outcomes" to make estimates of properties like Bayes' risk.

**Comment by ihincks** [comment]
(But yes, I will add to the guide)

**Comment by cgranade** [comment]
Ah, that makes a lot more sense, then. Thank you for the explanation, and for all the help!

File: src/qinfer/domains.py [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+        """
+        Returns ``True`` if all of the given points are in the domain, 
+        ``False`` otherwise.
+
+        :param np.ndarray points: An `np.ndarray` of type `self.dtype`.
+
+        :rtype: `bool`
+        """
+        array_view = self.to_regular_array(points)
+        print(np.all(array_view>=0))
```
Comment (scasagrande) [inline src/qinfer/domains.py]: If you must include a `print` call, which I don't usually suggest, be sure to to include `from __future__ import print_function`

**Comment by scasagrande** [comment]
Py2.7 is having a problem due to the `print` statement that I have commented on. Conda 3.5 has an issue with the doctests.

As a reminder though for everyone, a PR of this magnitude is impossible to code review (and thus, I haven't looked over the actual implementation). Future work should be done in as small of chunks as possible. It keeps thing reviewable, reduces churn, and helps prevent the problem where peoples' local code branches too heavily from master, creating massive merge conflicts.

**Comment by ihincks** [comment]
Thanks for catching that forgotten print statement @scasagrande.

I agree that this PR is big. At the same time, it really does affect all of these files: if you add some methods and properties to a fundamental abstract class, you gotta fix everything that inherits from it. And then I introduced more than 100 tests involving all inheriting classes to try to minimize errors.

**Comment by scasagrande** [comment]
There is always ways to split the work up, and part of learning to work on a code base with a team is submitting code-review-able PRs. This might involve a number of PRs merged into a staging branch before merging into a primary branch.

Some ideas on how to slice large base changes like this:

- Whitespace. If your project doesn't have a set of linting rules you'll probably have whitespace changes. If you know your work going to result in a lot of line changes, get this out of the way first so as to not overwhelm people with green lines.
- Tests for existing functionality / APIs. These tests can be useful even if you don't make your other functionality changes, so go ahead and make them a separate PR
- Base class changes, and minimum class changes. I'd skip class-specific implementations at this point. This can go into a staging branch.
- Class specific implementations. Whether this is one PR or one for each class depends on how intense the changes are. For this work, probably just one.

To give you an idea, at work, the worst case we had was when we had to rewrite this massive file. It was pure refactoring work, but had to be done so that our actual work could be implemented (so one degree of PR-splitting already). We had to change it from being some all-knowing god-object into using proper dependency injection. We split it as much as we could (one PR for new tests, one for the refactoring) but it was still ugly to review.

**Comment by ihincks** [comment]
Okay @cgranade, I have modified the ``models.rst`` guide to have a section on outcomes, and to not fail due to ``Model`` vs ``FiniteOutcomeModel``. I also have included some more API docs for Domains, MultinomialModel, etc.

**Comment by cgranade** [comment]
Wonderful, thank you so much @ihincks! I've looked over the documentation, and it all builds fine, looks great, so thank you as well for adding that. Since it all passes now, I think it's ready to merge in. Thank you again!

**Pull request merged**